compiler/src/middle_end/matchcomp.re

```diff
@@ -914,156 +667,503 @@ let rec compile_matrix = mtx =>
+      let env = expr.imm_env;
+      (
+        Comp.imm(
+          ~allocation_type=StackAllocated(WasmI32),
+          Imm.const(Const_number(Const_number_int(Int64.of_int(i)))),
+        ),
+        get_bindings(~mut_boxing, env, patterns, values, aliases),
+      );
+    | Guard(mb, patterns, aliases, true_tree, false_tree) =>
+      let guard =
+        switch (mb.mb_guard) {
+        | Some(guard) => guard
+        | None =>
+          failwith(
+            "Impossible: (compile_tree_help): guarded match branch contained no guard",
+          )
+        };
+      let bindings =
+        get_bindings(
+          ~mut_boxing,
+          mb.mb_body.exp_env,
+          patterns,
+          values,
+          aliases,
+        );
+      let (cond, cond_setup) = helpI(guard);
+      let (true_comp, true_setup) =
+        compile_tree_help(true_tree, values, expr, helpI, helpConst);
+      let (false_comp, false_setup) =
+        compile_tree_help(false_tree, values, expr, helpI, helpConst);
+      (
+        Comp.if_(
+          ~allocation_type=true_comp.comp_allocation_type,
+          cond,
+          fold_tree(true_setup, true_comp),
+          fold_tree(false_setup, false_comp),
+        ),
+        bindings @ cond_setup,
+      );
+    | Conditional((equality_type, const), alias, true_tree, false_tree) =>
+      let (cur_value, rest_values) =
+        switch (values) {
+        | [] => failwith("Impossible (compile_tree_help): Empty value stack")
+        | [hd, ...tl] => (hd, tl)
+        };
+      let equality_op =
+        switch (equality_type) {
+        | StructuralEquality => Eq
+        | PhysicalEquality => Is
+        };
+      let (const, const_setup) =
+        switch (helpConst(const)) {
+        | Left(imm) => (imm, [])
+        | Right((name, comp)) =>
+          let id = Ident.create(name);
+          (Imm.id(id), [BLet(id, comp, Nonglobal)]);
+        };
+      let cond =
+        Comp.prim2(
+          ~allocation_type=StackAllocated(WasmI32),
+          equality_op,
+          cur_value,
+          const,
+        );
+      let cond_id = Ident.create("match_conditional");
+      let cond_setup = const_setup @ [BLet(cond_id, cond, Nonglobal)];
+      let (true_comp, true_setup) =
+        compile_tree_help(true_tree, values, expr, helpI, helpConst);
+      let (false_comp, false_setup) =
+        compile_tree_help(false_tree, values, expr, helpI, helpConst);
+
+      // Add a binding for aliases
+      let alias_binding =
+        BLet(
+          alias,
+          Comp.imm(~allocation_type=HeapAllocated, cur_value),
+          Nonglobal,
+        );
+
+      (
+        Comp.if_(
+          ~allocation_type=true_comp.comp_allocation_type,
+          Imm.id(cond_id),
+          fold_tree(true_setup, true_comp),
+          fold_tree(false_setup, false_comp),
+        ),
+        [alias_binding, ...cond_setup],
+      );
+    | Fail =>
+      /* FIXME: We need a "throw error" node in ANF */
```

Comment:
Fix when? Do we need it in this PR or an issue?

Reply:
This was just moved here but it was outdated anyway, `Imm.trap()` is quite literally a throw error node haha